src/equations/serre_green_naghdi_flat_1d.jl

```diff
@@ -0,0 +1,456 @@
+                      solver,
+                      initial_condition,
+                      ::BoundaryConditionPeriodic,
+                      RealT, uEltype)
+    D = solver.D1
+
+    # create temporary storage
+    h = ones(RealT, nnodes(mesh))
+    h_x = zero(h)
+    v_x = zero(h)
+    h2_x = zero(h)
+    hv_x = zero(h)
+    v2_x = zero(h)
+    h2_v_vx_x = zero(h)
+    h_vx_x = zero(h)
+    p_x = zero(h)
+    tmp = zero(h)
+    M_h = zero(h)
+    M_h3_3 = zero(h)
+
+    if D isa PeriodicUpwindOperators
+        v_x_upwind = zero(h)
+
+        Dmat_minus = sparse(D.minus)
+
+        # Floating point errors accumulate a bit and the system matrix
+        # is not necessarily perfectly symmetric but only up to
+        # round-off errors. We wrap it here to avoid issues with the
+        # factorization.
+        @. M_h = h
+        scale_by_mass_matrix!(M_h, D)
+        @. M_h3_3 = (1 / 3) * h^3
+        scale_by_mass_matrix!(M_h3_3, D)
+        system_matrix = Symmetric(Diagonal(M_h)
+                                  +
+                                  Dmat_minus' * Diagonal(M_h3_3) * Dmat_minus)
+        factorization = cholesky(system_matrix)
+
+        cache = (; h_x, v_x, v_x_upwind, h2_x, hv_x, v2_x,
+                 h2_v_vx_x, h_vx_x, p_x, tmp,
+                 M_h, M_h3_3,
+                 D, Dmat_minus, factorization)
+    else
+        if D isa FourierDerivativeOperator
+            Dmat = Matrix(D)
+
+            cache = (; h_x, v_x, h2_x, hv_x, v2_x,
+                     h2_v_vx_x, h_vx_x, p_x, tmp,
+                     M_h, M_h3_3,
+                     D, Dmat)
+        else
+            Dmat = sparse(D)
+
+            # Floating point errors accumulate a bit and the system matrix
+            # is not necessarily perfectly symmetric but only up to
+            # round-off errors. We wrap it here to avoid issues with the
+            # factorization.
+            @. M_h = h
+            scale_by_mass_matrix!(M_h, D)
+            @. M_h3_3 = (1 / 3) * h^3
+            scale_by_mass_matrix!(M_h3_3, D)
+            system_matrix = Symmetric(Diagonal(M_h)
+                                      +
+                                      Dmat' * Diagonal(M_h3_3) * Dmat)
+            factorization = cholesky(system_matrix)
+
+            cache = (; h_x, v_x, h2_x, hv_x, v2_x,
+                     h2_v_vx_x, h_vx_x, p_x, tmp,
+                     M_h, M_h3_3,
+                     D, Dmat, factorization)
+        end
+    end
+
+    return cache
+end
+
+# Discretization that conserves
+# - the total water mass (integral of h) as a linear invariant
+# - the total momentum (integral of h v) as a nonlinear invariant
+# - the total energy
+# for periodic boundary conditions, see
+# - Hendrik Ranocha and Mario Ricchiuto (2024)
+#   Structure-preserving approximations of the Serre-Green-Naghdi
+#   equations in standard and hyperbolic form
+#   [arXiv: 2408.02665](https://arxiv.org/abs/2408.02665)
+# TODO: Implement source terms
```

Comment:
Yes, that should work. However, I need to derive a meaningful test first. Thus, I would like to postpone this - since this PR is already quite big. I have added it to the list in #129